### PR TITLE
B 4/create service provider entity

### PR DESCRIPTION
**Pull request by JulianFloresDev**
image

image
image

image
image
image
image
image

### CODE REVIEW

File: src/models/ServiceProviders.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import mongoose from "mongoose";
+
+const { Schema } = mongoose;
```
Comment (DamianPalavecino) [inline src/models/ServiceProviders.js]:
```suggestion
import mongoose, { Schema } from "mongoose";
```

File: src/index.js [hunk 2]
```diff
@@ -8,10 +8,11 @@ dotenv.config();
 
 const app = express();
 app.use(cors());
+app.use(express.json());
```
Comment (DamianPalavecino) [inline src/index.js]:
```suggestion
app.use(express.json());
app.use(cors());
```

File: src/controllers/serviceproviders.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  } catch (error) {
+    res.status(500).json({
+      message: `Unespected error ${error}.`,
+      error: true,
+      data: undefined,
+    });
+  }
+};
+
+export const getOneProvider = async (req, res) => {
```
Comment (DamianPalavecino) [inline src/controllers/serviceproviders.js]:
```suggestion
export const getProviderById = async (req, res) => {
```

File: src/controllers/serviceproviders.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      });
+    }
+    await newProvider.save();
+    return res.status(201).json({
+      message: `Service provider created successfully`,
+      data: newProvider,
+      error: false,
+    });
+  } catch (error) {
+    res.status(500).json({
```
Comment (DamianPalavecino) [inline src/controllers/serviceproviders.js:29]:
```suggestion
    res.status(400).json({
```

File: src/controllers/serviceproviders.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      message: `Unespected error ${error}.`,
```
Comment (DamianPalavecino) [inline src/controllers/serviceproviders.js]:
```suggestion
      message: `Unexpected error: ${error}.`,
```

Comment (DamianPalavecino) [reply src/controllers/serviceproviders.js]: Fix the rest of the messages

File: src/controllers/serviceproviders.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      error: true,
+      data: undefined,
+    });
+  }
+};
+
+export const getOneProvider = async (req, res) => {
+  try {
+    const { id } = req.body;
+    const provider = await ServiceProviders.findById(id).populate(
```
Comment (DamianPalavecino) [inline src/controllers/serviceproviders.js:73]: Here you can add a validation that checks that the id sent as path params is a valid id using: 

`function isValidObjectId(stringToCheck) {
    return mongoose.Types.ObjectId.isValid(stringToCheck);
}`

File: src/controllers/serviceproviders.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    const newProvider = await ServiceProviders.create({
+      ...req.body,
+      isActive: true,
+    });
+    const isProviderRegister = providers.some(
+      (prov) => prov.email === newProvider.email
+    );
+    if (isProviderRegister) {
+      return res.status(451).json({
+        message: "This email has already register",
```
Comment (DamianPalavecino) [inline src/controllers/serviceproviders.js]:
```suggestion
        message: "This email has already been registered",
```

File: src/controllers/serviceproviders.js (hunk 6 above)
Comment (DamianPalavecino) [reply src/controllers/serviceproviders.js:73]: Same in UpdateProvider

**Review by DamianPalavecino — APPROVED** [review]
Approved, good job

**Review by DamianPalavecino — APPROVED** [review]

**Review by Eju97 — APPROVED** [review]

**Pull request merged**